cmake/compilers/GNU.cmake

```diff
@@ -35,6 +35,40 @@ if (NOT CMAKE_GENERATOR MATCHES "Ninja" AND NOT CMAKE_CXX_DEPENDS_USE_COMPILER)
     set(TBB_MMD_FLAG -MMD)
 endif()
 
+
+# Binutils < 2.31.1 do not support the tpause instruction. When compiling with
+# a modern version of GCC (supporting it) but relying on an outdated assembler,
+# will result in an error reporting "no such instruction: tpause".
+# The following code invokes the GNU assembler to extract the version number
+# and convert it to an integer that can be used in the C++ code to compare
+# against, and conditionally disable the __TBB_WAITPKG_INTRINSICS_PRESENT
+# macro if the version is incompatible. Binutils only report the version in the
+# MAJOR.MINOR format, therefore the version checked is >=2.32 (instead of
+# >=2.31.1). Capturing the output in CMake can be done like below. The version
+# information is written to either stdout or stderr. To not make any
+# assumptions, both are captured.
+execute_process(
+    COMMAND ${CMAKE_CXX_COMPILER} -xc -c /dev/null -Wa,-v -o/dev/null
+    OUTPUT_VARIABLE ASSEMBLER_VERSION_LINE_OUT
+    ERROR_VARIABLE ASSEMBLER_VERSION_LINE_ERR
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+    ERROR_STRIP_TRAILING_WHITESPACE
+)
+set(ASSEMBLER_VERSION_LINE ${ASSEMBLER_VERSION_LINE_OUT}${ASSEMBLER_VERSION_LINE_ERR})
+string(REGEX REPLACE "GNU assembler version ([0-9]+\\.[0-9]+).*" "\\1" GNU_ASSEMBLER_VERSION "${ASSEMBLER_VERSION_LINE}")
+string(FIND ${GNU_ASSEMBLER_VERSION} "." DOT_POSITION)
+if(DOT_POSITION EQUAL -1)
+    message(FATAL_ERROR "Invalid version string format: ${GNU_ASSEMBLER_VERSION}")
+endif()
+
+# Extract the major and minor version numbers (without relying on REGEX to support older CMake versions <3.9)
+string(SUBSTRING ${GNU_ASSEMBLER_VERSION} 0 ${DOT_POSITION} GNU_ASSEMBLER_MAJOR_VERSION)
+string(SUBSTRING ${GNU_ASSEMBLER_VERSION} ${DOT_POSITION} -1 DOT_AND_MINOR_VERSION)
+string(SUBSTRING ${DOT_AND_MINOR_VERSION} 1 -1 GNU_ASSEMBLER_MINOR_VERSION)
```

Comment:
Why can't we just use two `string(REGEX REPLACE...` commands (it is supported in CMake 3.5)?
Like this:
```cmake
string(REGEX REPLACE "GNU assembler version ([0-9]+)\\.([0-9]+).*" "\\1" GNU_ASM_MAJOR_VERSION "${ASSEMBLER_VERSION_LINE}")
string(REGEX REPLACE "GNU assembler version ([0-9]+)\\.([0-9]+).*" "\\2" GNU_ASM_MINOR_VERSION "${ASSEMBLER_VERSION_LINE}")
```

Reply:
Clever. :)